crates/typst-library/src/layout/transform.rs

```diff
@@ -172,16 +239,71 @@ impl Layout for ScaleElem {
         styles: StyleChain,
         regions: Regions,
     ) -> SourceResult<Fragment> {
-        let pod = Regions::one(regions.base(), Axes::splat(false));
+        let sx = self.x(styles);
+        let sy = self.y(styles);
+        let layout = self.reflow(styles);
+        let align = self.origin(styles).resolve(styles);
+
+        if !layout {
+            // Layout the body.
+            let pod = Regions::one(regions.base(), Axes::splat(false));
+            let mut frame = self.body().layout(vt, styles, pod)?.into_frame();
+            let Axes { x, y } = self
+                .origin(styles)
+                .resolve(styles)
+                .zip_map(frame.size(), FixedAlign::position);
+
+            // Apply the transform.
+            let ts = Transform::translate(x, y)
+                .pre_concat(Transform::scale(sx, sy))
+                .pre_concat(Transform::translate(-x, -y));
+            frame.transform(ts);
+
+            return Ok(Fragment::frame(frame));
+        }
+
+        // Compute the new region's approximate size.
+        let size = regions.base().zip_map(Axes::new(sx, sy), |r, s| s.of(r));
+
+        // Measure the size of the body.
+        let pod = Regions::one(size.map(Abs::abs), Axes::splat(false));
+        let frame = self.body().measure(vt, styles, pod)?.into_frame();
+
+        // Actually perform the layout.
+        let pod = Regions::one(frame.size(), Axes::splat(true));
         let mut frame = self.body().layout(vt, styles, pod)?.into_frame();
-        let Axes { x, y } = self
-            .origin(styles)
-            .resolve(styles)
-            .zip_map(frame.size(), FixedAlign::position);
-        let transform = Transform::translate(x, y)
-            .pre_concat(Transform::scale(self.x(styles), self.y(styles)))
+        let Axes { x, y } = align.zip_map(frame.size(), FixedAlign::position);
+
+        // Apply the transform.
```

Comment:
I feel like there is quite a lot of code duplication in the two code paths. Would it be possibly to unify them?